### PR TITLE
New function `pureIf`

### PR DESCRIPTION
**Pull request by mel-brown**
These are functions I wish were in `base`. How many times have I written `if b then Just a else Nothing`...?

Instead I want to write `a <? b` or `b ?> a` or ```a `justIf` b```.

The infixes bind lower than `(&&)` and `(||)` so that they can be used inside the `Bool` argument with fewer parentheses.

### CODE REVIEW

**Comment by mel-brown** [comment]
Can't figure out how to make the CI tests work, it looks like scope errors.

**Comment by ndmitchell** [comment]
Thanks for the patch. Let's not worry about the tests just yet (I think you just need to add it to TestUtil.hs), but the functions themselves (I'm not whatever has upset the tests is easily fixable). I too have done the justIf pattern a lot, and typically do it with:

```
guard bool
pure result
```

So I think encoding that as `justIf` makes sense. The remaining things to think about are name, type signature and argument order. My inclination is the argument order would be the other way round - since its encoding `if bool then Just result else Nothing`. For type, we could generalise it to Alternative, but I'm quite happy with the more concrete `Maybe` restricted variant, as I imagine that's the commonly used one. For name, `justIf` or `justWhen` are the ones I can think of, and `justIf` seems better - although I wouldn't be expecting to use it infix, but prefix usually (although that's just style).

As for defining operators for it, currently extra tries to only define things where when people see the name its obvious. As a result, extra shies away from operators - only things like `&&^` which are "relatively" obvious extensions of `&&`. I'd be inclined to add `justIf` but without operators. Thoughts?

**Comment by josephcsible** [comment]
I agree we should put the `Bool` first. I'm also in favor of generalizing this to arbitrary `Alternative`s.

**Comment by mel-brown** [comment]
Sounds good then. You make a fair argument about the operators, so I'll remove those. I think `justIf` generalizes nicely to `pureIf`, so I'll toss that one in `Control.Alternative.Extra`. And, I'll put the Boolean first, so that it lines up with other functions like `when` and `unless`.

File: src/Control/Applicative/Extra.hs [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+module Control.Applicative.Extra (
+    module Control.Applicative,
+    pureIf,
+) where
+
+import Control.Applicative
+
+-- | A generalized version of 'Data.Bool.Extra.justIf'.
+pureIf :: (Alternative m) => Bool -> a -> m a
+pureIf b a = if b then pure a else empty
```
Comment (ndmitchell) [inline src/Control/Applicative/Extra.hs:11]: Can we include a trailing newline in this file.

File: src/Data/Bool/Extra.hs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+) where
+
+import Data.Bool
+
+-- | Return either the value wrapped in a `Just`, or `Nothing` if the condition is false.
+--
+-- > justIf True  5 == Just 5
+-- > justIf False 5 == Nothing
+justIf :: Bool -> a -> Maybe a
+justIf b a = if b then Just a else Nothing
```
Comment (ndmitchell) [inline src/Data/Bool/Extra.hs:15]: Trailing newline again (otherwise it shows up in editors and diffs)

**Review by ndmitchell — COMMENTED** [review]
Looks good. My only question is where two new modules it the right place for these to go. I'd be inclined to put pureIf in control.monad (which is broadly things that participate in the Monad heirarchy - lots are applicative, and this particular one is Alternative rather than applicative anyway). I'd also be tempted to put `justIf` in Data.Maybe.Extra`, as its more about `Just` than it is about Bool. Thoughts?

**Comment by mel-brown** [comment]
It looks like `Control.Monad.Extra` already has `whenMaybe`, which is extremely similar to `justIf`:

```haskell
whenMaybe :: Applicative m => Bool -> m a -> m (Maybe a)
whenMaybe b x = if b then Just <$> x else pure Nothing
```

I think this is a good home for `pureIf`. But seeing as there is no existing module `Data.Maybe.Extra` I think `justIf` is just as well where it is, since it can be viewed as a specialized eliminator for `Bool`.

**Comment by ndmitchell** [comment]
Sorry, I've been underwater with email for a few weeks, just recovering.

Looking at things as they now stand, I have two questions:

1. Does `justIf` carry its own weight? E.g. is `pureIf` meaningfully worse? I'm about 50:50 on this, so happy to go with whatever you think.
2. If `justIf` stays, should it really be in Data.Maybe.Extra. You've taken `pureIf` and specialised it to `Maybe`, so having it in Data.Bool.Extra doesn't seem right.

And can `pureIf` get some docs beyond just a generalisation of `justIf` - e.g.

```
-- > pureIf True 1 == Just 1
-- > pureIf False 1 == Nothing
-- > pureIf True 1 == [1]
-- > pureIf False 1 == []
```

Sorry for all these revisions - it's looking really close now.

**Comment by josephcsible** [comment]
> It looks like `Control.Monad.Extra` already has `whenMaybe`, which is extremely similar to `justIf`

For the record, they're related in terms of `sequenceA`: `whenMaybe b = sequenceA . justIf b`. (And in fact that reveals there could be `whenAlt b = sequenceA . pureIf b` too, but I'll wait to propose that until I actually come across a use case for it.)

> Does `justIf` carry its own weight? E.g. is `pureIf` meaningfully worse? I'm about 50:50 on this, so happy to go with whatever you think.

My $0.02: no, so we should get rid of `justIf` and only add `pureIf`.

**Comment by mel-brown** [comment]
I agree, `pureIf` by itself makes the most sense. I'll keep it in `Control.Monad.Extra` and include in the docs the `Maybe` and `[]` examples.

**Comment by mel-brown** [comment]
It looks like the doctests are failing because I put type annotations in for clarity in the Haddocks. What’s the best solution?

**Comment by ndmitchell** [comment]
Thanks! The test code generator needs adjusting - I'll do that myself in a follow up. Thanks for the contribution!

**Comment by ndmitchell** [comment]
All done in the commits above, so CI should now be happy.

**Pull request merged**